common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java

```diff
@@ -0,0 +1,258 @@
+package gov.cms.ab2d.common.service;
+
+import gov.cms.ab2d.common.model.Coverage;
+import gov.cms.ab2d.common.model.CoveragePeriod;
+import gov.cms.ab2d.common.model.CoverageSearchDiff;
+import gov.cms.ab2d.common.model.CoverageSearchEvent;
+import gov.cms.ab2d.common.model.JobStatus;
+import gov.cms.ab2d.common.repository.CoveragePeriodRepository;
+import gov.cms.ab2d.common.repository.CoverageRepository;
+import gov.cms.ab2d.common.repository.CoverageSearchEventRepository;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.Collection;
+import java.util.Optional;
+
+import static gov.cms.ab2d.common.util.Constants.AB2D_EPOCH;
+import static java.util.stream.Collectors.toList;
+
+@Slf4j
+@Service
+@Transactional
+public class CoverageServiceImpl implements CoverageService {
+
+    @Autowired
+    private CoverageRepository coverageRepo;
+
+    @Autowired
+    private CoveragePeriodRepository coveragePeriodRepo;
+
+    @Autowired
+    private CoverageSearchEventRepository coverageSearchEventRepo;
+
+    @Override
+    public CoveragePeriod getCoveragePeriod(long contractId, int month, int year) {
+        checkMonthAndYear(month, year);
+        return coveragePeriodRepo.getByContractIdAndMonthAndYear(contractId, month, year);
+    }
+
+    @Override
+    public boolean isCoveragePeriodInProgress(int searchId) {
+        CoveragePeriod period = coveragePeriodRepo.getOne(searchId);
+        JobStatus jobStatus = period.getStatus();
+        return jobStatus == JobStatus.IN_PROGRESS;
+    }
+
+    @Override
+    public boolean canEOBSearchBeStarted(int searchId) {
+        return !isCoveragePeriodInProgress(searchId);
+    }
+
+    @Override
+    public JobStatus getSearchStatus(int searchId) {
+
+        CoveragePeriod coverageSearch = coveragePeriodRepo.getOne(searchId);
+        return coverageSearch.getStatus();
+
+    }
+
+    @Override
+    public Optional<CoverageSearchEvent> findLastEvent(int searchId) {
+        CoveragePeriod period = coveragePeriodRepo.getOne(searchId);
+        return findLastEvent(period);
+    }
+
+    private Optional<CoverageSearchEvent> findLastEvent(CoveragePeriod period) {
+        return coverageSearchEventRepo.findFirstByCoveragePeriodOrderByCreatedDesc(period);
+    }
+
+    private CoverageSearchEvent getLastEvent(CoveragePeriod period) {
+        return findLastEvent(period).orElse(null);
+    }
+
+    @Transactional
+    @Override
+    public CoverageSearchEvent insertCoverage(int searchId, long searchEventId, Collection<String> beneficiaryIds) {
```

Comment:
Not sure you need transactional at the method level in all these places if it's at the class level.